Update validation page. (#1803)
graphql/graphql.github.io: src/pages/learn/validation.mdx

# Validation

<p className="learn-subtitle">Learn how GraphQL validates operations using a schema</p>

On this page, we'll explore an important phase in the lifecycle of a GraphQL request called [validation](https://spec.graphql.org/draft/#sec-Validation). A request must be syntactically correct to run, but it should also be valid when checked against the API's schema. 

> In practice, when a GraphQL operation reaches the server, the document is first parsed and then validated using the type system. This allows servers and clients to effectively inform developers when an invalid query has been created, and without relying on runtime checks. Once the operation is validated, it can be [executed](/learn/execution/) on the server and a response will be delivered to the client.

Reviewer: ```suggestion
In practice, when a GraphQL operation reaches the server, the document is first parsed and then validated using the type system. This allows servers and clients to effectively inform developers when an invalid query has been created, without relying on runtime checks. Once the operation is validated, it can be [executed](/learn/execution/) on the server and a response will be delivered to the client.
```

## Validation examples

> The GraphQL specification describes the detailed conditions that must be satisfied for a request to be considered valid. In the sections that follow, we'll look at a few examples of common validation issues that occur in GraphQL operations.

Reviewer: ```suggestion
The GraphQL specification describes the detailed conditions that must be satisfied for a request to be considered valid. In the sections that follow, we'll look at a few examples of common validation issues that may occur in GraphQL operations.
```

### Requesting non-existent fields

> When we query for fields, we have to request a field on the given type. So as `hero` returns a `Character` type, we have to query for a field that's defined on `Character`. That type does not have a `favoriteSpaceship` field, so this query is invalid:

Reviewer: I recognise that this was the original text, but it doesn't read well to me. Here's a proposed alternative, what do you think?

```suggestion
When we query for a field, the field must be defined on the relevant type. As `hero` returns a `Character` type, its selection set may only request the `Character` type's fields; `Character` does not have a `favoriteSpaceship` field, so this query is invalid:
```

Incidentally it would be much nicer to write `Query.hero` here, but to do so we'd have to introduce "schema coordinates" earlier.

```graphql
# { "graphiql": true }
# INVALID: favoriteSpaceship does not exist on Character

> query {

Reviewer: Interested to know why you're moving away from the query shorthand here? It tends to be very rare that I see this - either the operation has a name and/or variables in which case you need `query`, or it has neither in which case it's very rare to see it in my experience.

Author: I've tended to see the opposite (inclusion of the `query` keyword) so I included it for clarity after all of the keywords are introduced on the operation pages. But I can remove it throughout if the shorthand syntax is the preferred convention.

Reviewer: Let's worry about that afterwards, if at all; it's much easier to replace `/^query {$/` with `{` than the reverse - so keep going with the `query` keyword for now.

Interesting that your experience differs; I wonder what the ecosystem as a whole sees!

  hero {
    favoriteSpaceship
  }
}
```

> ### Selecting invalid leaf fields

Reviewer: Perhaps this is a clearer title?

```suggestion
### Selection sets and leaf fields
```

> Whenever we query for a field and it returns something other than a Scalar or Enum type, we need to specify what data we want to get back from the field. The `hero` query field returns a `Character`, and we've already seen examples that request fields like `name` and `appearsIn` on it. If we omit those leaf field selections, then the query will not be valid:

Reviewer: ```suggestion
Whenever we query for a field and it returns something other than a Scalar or Enum type, we need to specify what data we want to get back from the field (a "selection set"). The `hero` query field returns a `Character`, and we've already seen examples that request fields like `name` and `appearsIn` on it. If we omit those leaf field selections, then the query will not be valid:
```

```graphql
# { "graphiql": true }
# INVALID: hero is not a scalar, so fields are needed
query {
  hero
}
```

> Similarly, if the leaf field is a scalar value, it doesn't make sense to query for additional fields on it, and doing so will make the query invalid:

Reviewer: This statement is true of leaf fields whether or not they are scalar values (also applies to enums); maybe this edit is clearer?

```suggestion
Similarly, querying fields of a scalar or enum doesn’t make sense, therefore adding a selection set to a leaf field will make the query invalid:
```

Alternatively maybe something like "Similarly, since a leaf field represents a scalar or enum, it doesn't make sense..."?

}
```

### Cyclic fragment spreads

To start, let's take a complex valid query. This is a nested query, but with the duplicated fields factored out into a fragment:

```graphql
# { "graphiql": true }
query {
  hero {
    ...NameAndAppearances
    friends {
      ...NameAndAppearances
      friends {
        ...NameAndAppearances
      }
    }
  }
}

fragment NameAndAppearances on Character {
  name
  appearsIn
}
```

> And this query is valid. Let's take a look at some invalid queries...
> 
> A fragment cannot refer to itself or create a cycle, as this could result in an unbounded result! Here's the same query above but without the explicit three levels of nesting:

Reviewer: We mentioned the query was valid already and we only have one query below here; here's an attempt to merge these paragraphs:

```suggestion
The following is an alternative to the above query, attempting to use recursion instead of the explicit three levels of nesting. This new query is invalid because a fragment cannot refer to itself (directly or indirectly) since the resulting cycle could create an unbounded result!
```

```graphql
# { "graphiql": true }
query {
  hero {
    ...NameAndAppearancesAndFriends
  }
}

fragment NameAndAppearancesAndFriends on Character {
  name
  appearsIn
  friends {
    ...NameAndAppearancesAndFriends
  }
}
```

> This has just scratched the surface of the validation system; there are a number of validation rules in place to ensure that a GraphQL operation is semantically meaningful. The specification goes into more detail about this topic in the [validation section](https://spec.graphql.org/draft/#sec-Validation), and the [validation directory](https://github.com/graphql/graphql-js/blob/main/src/validation) in GraphQL.js contains code implementing a specification-compliant GraphQL validator.

Reviewer: ```suggestion
This has just scratched the surface of the validation system; there are a number of validation rules in place to ensure that a GraphQL operation is semantically meaningful. The specification goes into more detail about this topic in the [validation section](https://spec.graphql.org/draft/#sec-Validation), and the [validation directory in the reference implementation](https://github.com/graphql/graphql-js/blob/main/src/validation) contains code implementing a specification-compliant GraphQL validator.
```

## Validation errors

As we have seen in the examples above, when a GraphQL server encounters a validation error in a request, it will return information about what happened in the `errors` key of the response. Specifically, when GraphQL detects a validation issue in a document, it raises a _request error_ before execution begins, meaning that no partial data will be included in the response.

And because the GraphQL specification requires all implementations to validate incoming requests against the schema, developers won't need to write specific runtime logic to handle these validation issues manually.

## Next steps

To recap what we've learned about validation:

- The GraphQL documents that clients submit in their requests must be syntactically correct and considered valid when checked against the schema
- The GraphQL specification requires that implementations check that incoming requests contain valid field selections, correct fragment usage, and more

> - When a validation issue occurs, the server will raise a request error and return information about what happened to the client before field execution occurs

Reviewer: ```suggestion
- To be executed, requests must include a syntactically correct document that is considered valid when checked against the schema
- The specification requires implementations check incoming requests contain valid field selections, correct fragment usage, and more
- When a validation issue occurs, the server will raise a request error and return to the client information about what happened; field execution will not start
```

> Head over to the [Execution](/learn/execution/) page to learn how GraphQL provides data for each field in a request after the validation step is complete.

Reviewer: ```suggestion
Head over to the [Execution](/learn/execution/) page to learn how GraphQL provides data for each field in a request after the validation step successfully completes.
```
